File: LEETCODE/LEETCODE-CPP/206.cpp
//206. Reverse Linked List
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) 
    {
        struct ListNode *q=NULL;
        struct ListNode *p;
        struct ListNode *t=head;
        if(head==NULL || head->next==NULL)
            return head;
        else
           while(t) 
           {
               p=q;
               q=t;
               t=t->next;
               q->next=p;
           }
        head=q;
        return head;
    }
};

File: LEETCODE/LEETCODE-PYTHON/206.py
# 206. Reverse Linked List
# https://leetcode.com/problems/reverse-linked-list/description/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev=None
        curr=head

        while curr:
            next_temp=curr.next
            curr.next=prev
            prev=curr
            curr=next_temp

        return prev